File: src/adeu/mcp_components/_response_builders.py
# FILE: src/adeu/mcp_components/_response_builders.py
"""
Shared response builders for read_docx mode dispatch.

Lives in mcp_components (not in tools/) because both tools/document.py and
tools/live_word.py call into it. Keeping it here avoids the circular import
that would result if these helpers lived in document.py and live_word.py
imported them at module load time.

PARITY INVARIANT
----------------
Both the disk path (_read_docx_disk) and the Live Word path
(read_active_word_document) MUST converge on these builders for outline and
pagination output. The contract is:

  Given the same DOCX content (whether read from disk or extracted via
  WordOpenXML from a saved-state Live Word doc), build_paginated_response and
  build_outline_response MUST return byte-identical results.

This invariant is what justifies the design — there is no Live-Word-specific
pagination or outline code. Any future feature that touches projection,
pagination, or outline should be added to ingest/pagination/outline so both
paths inherit the change automatically.

CHANNEL CONTRACT
----------------
Per the MCP spec, `content` is LLM-facing markdown and `structured_content` is
machine-facing JSON for the host UI. We do NOT mirror them; instead we ensure
each channel is self-sufficient for its audience:

  - `content`: contains the projected document text PLUS an inline pagination
    banner (top, and bottom-of-page when has_next) so the LLM knows its
    position in the document without consulting structured_content.

  - `structured_content`: contains only fields the markdown UI widget actually
    reads — `markdown`, `title`, `file_path`. Everything else has been removed
    because nothing consumes it.

APPENDIX SEPARATION (Step 2)
----------------------------
The Structural Appendix (defined terms, anchors, diagnostics) is NOT included
in body pages. It is fetched on demand via mode='appendix'. Body pages get a
small one-line footer pointing the agent at the appendix mode. This was
necessary because on large legal documents the appendix can exceed 400KB,
which (a) blew the per-page payload ceiling and (b) was being silently
chunked by the MCP client.

The appendix is paginated using the same paginator as the body, with the
appendix text passed AS the body input.
"""

from pathlib import Path
from typing import TYPE_CHECKING, Any, List

from fastmcp.exceptions import ToolError
from fastmcp.tools.tool import ToolResult

from adeu.outline import extract_outline
from adeu.pagination import paginate, split_structural_appendix

if TYPE_CHECKING:
    from docx.document import Document as DocumentObject


def _build_appendix_pointer(file_path: str, has_appendix: bool) -> str:
    """
    Returns the one-line footer appended to body pages telling the agent that
    structural metadata (defined terms, cross-references, bookmarks, diagnostics)
    is available via mode='appendix'. Empty string when the document has no
    appendix content.

    This footer is the agent's signal that it should consult the appendix
    before making edits that touch defined terms or referenced sections.
    """
    if not has_appendix:
        return ""
    return (
        f"\n\n---\n\n"
        f"> **Appendix available.** This document has structural metadata "
        f"(defined terms, cross-references, bookmarks, diagnostics) that may "
        f"be relevant when editing. Call `read_docx` with `mode='appendix'` "
        f"to load it before submitting edits."
    )


def _build_page_banner(page: int, total: int) -> str:
    """
    Returns the top-of-page banner injected into LLM-facing markdown.
    Empty string when there is only one page (no navigation needed).
    """
    if total <= 1:
        return ""
    return (
        f"> **Page {page} of {total}** — "
        f"call `read_docx` with `mode='outline'` for a heading map of the full document.\n\n"
        f"---\n\n"
    )


def _build_page_footer(page: int, total: int, has_next: bool) -> str:
    """
    Returns the bottom-of-page continuation marker. Empty when this is the
    last page or the only page.
    """
    if total <= 1 or not has_next:
        return ""
    return f"\n\n---\n\n> **Continues on page {page + 1} of {total}.**"


def render_outline_tree(
    nodes: List[Any],
    max_level: int = 2,
    verbose: bool = False,
) -> str:
    """
    Renders a flat list of OutlineNode objects as a Markdown tree.

    Args:
        nodes: full list of OutlineNode objects from extract_outline().
        max_level: only render nodes at level <= max_level. Default 2 keeps
            the output usable on large documents (a 1000-page legal doc can
            have 7000+ Heading-4-styled paragraphs that drown out real
            navigation structure). Pass max_level=6 for full depth.
        verbose: when True, includes style name, has_table flag, and
            footnote IDs in the per-node metadata. Off by default to
            keep the payload small for the common navigation case.
    """
    if not nodes:
        return "# (No headings detected)\n\nThis document has no detectable headings."

    visible = [n for n in nodes if n.level <= max_level]

    if not visible:
        return (
            f"# (No headings at level <= {max_level})\n\n"
            f"Document has {len(nodes)} headings, all at deeper levels. "
            f"Call read_docx with mode='outline' and outline_max_level=N "
            f"(up to 6) to see them."
        )

    lines = []
    for node in visible:
        prefix = "#" * node.level
        if verbose:
            meta_parts = [f"p{node.page}", node.style]
            if node.has_table:
                meta_parts.append("has table")
            if node.footnote_ids:
                meta_parts.append("fn:" + ",".join(node.footnote_ids))
            meta = ", ".join(meta_parts)
            lines.append(f"{prefix} {node.text} ({meta})")
        else:
            lines.append(f"{prefix} {node.text} (p{node.page})")
    return "\n".join(lines)


def build_paginated_response(text: str, page: int, file_path: str) -> ToolResult:
    """
    Splits projected Markdown into pages and returns the requested page.

    The structural appendix is NOT included in the page content (since Step 2).
    Body pages get a one-line footer pointing the agent at mode='appendix'
    if the document has an appendix.

    Raises ToolError if `page` is out of range.
    """
    body, appendix = split_structural_appendix(text)
    has_appendix = bool(appendix.strip())

    # Paginate body only. Pass empty string as structural_appendix so the
    # paginator does not glue anything onto each page.
    result = paginate(body, structural_appendix="")

    if page < 1 or page > result.total_pages:
        raise ToolError(
            f"Page {page} out of range (doc has {result.total_pages} pages)."
        )

    selected = result.pages[page - 1]

    # Build the original UI markdown
    banner = _build_page_banner(selected.page, selected.total_pages)
    footer = _build_page_footer(selected.page, selected.total_pages, selected.has_next)
    appendix_pointer = _build_appendix_pointer(file_path, has_appendix)
    ui_markdown = banner + selected.page_content + footer + appendix_pointer

    # Prepend the path ONLY for the LLM
    llm_content = f"> **File Path:** `{file_path}`\n\n{ui_markdown}"

    return ToolResult(
        content=llm_content,
        structured_content={
            "markdown": ui_markdown,
            "title": Path(file_path).name,
            "file_path": str(Path(file_path).resolve()),
        },
    )


def build_outline_response(
    doc: "DocumentObject",
    projected_text: str,
    file_path: str,
    outline_max_level: int = 2,
    outline_verbose: bool = False,
    paragraph_offsets: dict | None = None,
) -> ToolResult:
    """
    Returns a structural map of headings as a Markdown tree.

    Args:
        outline_max_level: cap on heading depth shown (default 2).
        outline_verbose: include per-node style/table/footnote metadata.
        paragraph_offsets: when provided, enables the fast outline path that
            avoids re-projecting paragraphs. Caller obtains this from
            _extract_text_from_doc(return_paragraph_offsets=True).
    """

    # Pagination is used here only to compute body page boundaries for
    # heading->page mapping. We deliberately pass empty string instead of the
    # appendix because per-page appendix injection is gone (Step 2).
    body, _appendix = split_structural_appendix(projected_text)
    pagination_result = paginate(body, structural_appendix="")

    nodes = extract_outline(
        doc,
        body,
        pagination_result.body_pages,
        pagination_result.body_page_offsets,
        paragraph_offsets=paragraph_offsets,
    )
    rendered = render_outline_tree(
        nodes,
        max_level=outline_max_level,
        verbose=outline_verbose,
    )

    visible_count = sum(1 for n in nodes if n.level <= outline_max_level)
    deeper_count = len(nodes) - visible_count
    deeper_hint = (
        f" ({deeper_count} more at deeper levels, raise outline_max_level to see)"
        if deeper_count > 0
        else ""
    )

    # Build the original UI markdown
    header = (
        f"> **Outline view** — showing {visible_count} of {len(nodes)} headings "
        f"(L1-L{outline_max_level}{deeper_hint}) across "
        f"{pagination_result.total_pages} page(s). "
        f"Call `read_docx` with `mode='full'` and `page=N` to read a section.\n\n"
        f"---\n\n"
    )
    ui_markdown = header + rendered
    # Prepend the path ONLY for the LLM
    llm_content = f"> **File Path:** `{file_path}`\n\n{ui_markdown}"

    return ToolResult(
        content=llm_content,
        structured_content={
            "markdown": ui_markdown,
            "title": Path(file_path).name,
            "file_path": str(Path(file_path).resolve()),
        },
    )


def build_appendix_response(text: str, page: int, file_path: str) -> ToolResult:
    """
    Returns the structural appendix (defined terms, anchors, diagnostics) for
    the document, paginated. The appendix is treated AS the body for pagination
    purposes — same paginator, same boundary safety, same per-page banner.

    The agent fetches this on demand to inform editing decisions on documents
    where the body pages flag an appendix exists.

    Raises ToolError if `page` is out of range.
    Returns a single-page "no appendix" response if the document has no
    structural metadata.
    """
    _body, appendix = split_structural_appendix(text)

    if not appendix.strip():
        ui_markdown = (
            f"# Appendix\n\n"
            f"This document has no structural appendix "
            f"(no defined terms, named anchors, or diagnostics detected)."
        )
        llm_content = f"> **File Path:** `{file_path}`\n\n{ui_markdown}"
        return ToolResult(
            content=llm_content,
            structured_content={
                "markdown": ui_markdown,
                "title": Path(file_path).name,
                "file_path": str(Path(file_path).resolve()),
            },
        )

    # Treat the appendix AS the body and paginate it.
    result = paginate(appendix, structural_appendix="")

    if page < 1 or page > result.total_pages:
        raise ToolError(
            f"Appendix page {page} out of range (appendix has {result.total_pages} pages)."
        )

    selected = result.pages[page - 1]

    # Build the appendix-specific banner. Reusing _build_page_banner would emit
    # generic "Page N of M" wording; the agent benefits from knowing it's
    # looking at the appendix, not body.
    if selected.total_pages > 1:
        banner = (
            f"> **Appendix page {selected.page} of {selected.total_pages}** — "
            f"structural metadata for this document.\n\n---\n\n"
        )
        footer = (
            (
                f"\n\n---\n\n"
                f"> **Continues on appendix page {selected.page + 1} of "
                f"{selected.total_pages}.**"
            )
            if selected.has_next
            else ""
        )
    else:
        banner = "> **Appendix** — structural metadata for this document.\n\n---\n\n"
        footer = ""

    ui_markdown = banner + selected.page_content + footer
    llm_content = f"> **File Path:** `{file_path}`\n\n{ui_markdown}"

    return ToolResult(
        content=llm_content,
        structured_content={
            "markdown": ui_markdown,
            "title": Path(file_path).name,
            "file_path": str(Path(file_path).resolve()),
        },
    )

File: node/packages/mcp-server/src/response-builders.ts
import { resolve, basename } from "node:path";
import {
  DocumentObject,
  paginate,
  split_structural_appendix,
  extract_outline,
  OutlineNode,
  RegexTimeoutError,
  userFindAllMatches,
} from "@adeu/core";

export interface ToolResult {
  content: { type: "text"; text: string }[];
  structuredContent?: any;
  isError?: boolean;
  [key: string]: unknown;
}

// Projection style markers: `**bold**` always; `_italic_` only where the
// underscore is not intra-word (identifiers like snake_case are literal text —
// the projection's italics markers always hug non-whitespace at a word edge).
const STYLE_MARKER_RE = /\*\*|(?<![\w])_(?=\S)|(?<=\S)_(?![\w])/g;

/**
 * Renders `prefix **match** suffix` with the document's own bold/italic
 * projection markers stripped first, so the highlight cannot collide with
 * markers already present — a regex match crossing styled runs used to
 * render as `**The **Supplier** _shall provide**_` (QA 2026-07-19 v8 F-10).
 * Markers are detected over the WHOLE region (a match boundary can cut a
 * marker away from its word-edge context), then each part is rebuilt from
 * the surviving characters. Mirrors Python's _emphasized_snippet.
 */
export function emphasizedSnippet(
  prefix: string,
  match: string,
  suffix: string,
): string {
  const region = prefix + match + suffix;
  const b1 = prefix.length;
  const b2 = prefix.length + match.length;
  const keep = new Array<boolean>(region.length).fill(true);
  for (const m of region.matchAll(STYLE_MARKER_RE)) {
    for (let i = m.index!; i < m.index! + m[0].length; i++) keep[i] = false;
  }
  let strippedPrefix = "";
  let strippedMatch = "";
  let strippedSuffix = "";
  for (let i = 0; i < region.length; i++) {
    if (!keep[i]) continue;
    if (i < b1) strippedPrefix += region[i];
    else if (i < b2) strippedMatch += region[i];
    else strippedSuffix += region[i];
  }
  return `${strippedPrefix}**${strippedMatch}**${strippedSuffix}`;
}

function _build_appendix_pointer(has_appendix: boolean): string {
  if (!has_appendix) return "";
  return `\n\n---\n\n> **Appendix available.** This document has structural metadata (defined terms, cross-references, bookmarks, diagnostics) that may be relevant when editing. Call \`read_docx\` with \`mode='appendix'\` to load it before submitting edits.`;
}

function _build_page_banner(page: number, total: number): string {
  if (total <= 1) return "";
  // "synthetic" is load-bearing: Adeu pages are length-based content chunks
  // sized for LLM consumption, and readers must never mistake them for
  // printed Word pages or explicit page breaks (QA 2026-07-19 ADEU-QA-005).
  return `> **Page ${page} of ${total}** (synthetic page — a length-based chunk, not a printed Word page) — call \`read_docx\` with \`mode='outline'\` for a heading map of the full document.\n\n---\n\n`;
}

function _build_page_footer(
  page: number,
  total: number,
  has_next: boolean,
): string {
  if (total <= 1 || !has_next) return "";
  return `\n\n---\n\n> **Continues on page ${page + 1} of ${total}.**`;
}

export function render_outline_tree(
  nodes: OutlineNode[],
  max_level: number = 2,
  verbose: boolean = false,
): string {
  if (!nodes || nodes.length === 0) {
    return "# (No headings detected)\n\nThis document has no detectable headings.";
  }

  const visible = nodes.filter((n) => n.level <= max_level);

  if (visible.length === 0) {
    return `# (No headings at level <= ${max_level})\n\nDocument has ${nodes.length} headings, all at deeper levels. Call read_docx with mode='outline' and outline_max_level=N (up to 6) to see them.`;
  }

  const lines: string[] = [];
  for (const node of visible) {
    const prefix = "#".repeat(node.level);
    if (verbose) {
      const meta_parts = [`p${node.page}`, node.style];
      if (node.has_table) meta_parts.push("has table");
      if (node.footnote_ids && node.footnote_ids.length > 0)
        meta_parts.push("fn:" + node.footnote_ids.join(","));
      lines.push(`${prefix} ${node.text} (${meta_parts.join(", ")})`);
    } else {
      lines.push(`${prefix} ${node.text} (p${node.page})`);
    }
  }
  return lines.join("\n");
}

export function build_full_document_response(
  text: string,
  file_path: string,
): ToolResult {
  // The ENTIRE document body with no page banner, continuation footer, or
  // appendix pointer — the round-trip artifact for text-based apply/diff
  // (QA 2026-07-17 F1; mirrors Python's build_full_document_response).
  const [body] = split_structural_appendix(text);
  const ui_markdown = body;
  const llm_content = `> **File Path:** \`${resolve(file_path)}\`\n\n${ui_markdown}`;
  return {
    content: [{ type: "text", text: llm_content }],
    structuredContent: {
      markdown: ui_markdown,
      file_path: resolve(file_path),
      title: basename(file_path),
    },
  };
}

export function build_paginated_response(
  text: string,
  page: number,
  file_path: string,
): ToolResult {
  const [body, appendix] = split_structural_appendix(text);
  const has_appendix = Boolean(appendix.trim());

  const result = paginate(body, "");

  if (page < 1 || page > result.total_pages) {
    throw new Error(
      `Page ${page} out of range (doc has ${result.total_pages} pages).`,
    );
  }

  const selected = result.pages[page - 1];
  const banner = _build_page_banner(selected.page, selected.total_pages);
  const footer = _build_page_footer(
    selected.page,
    selected.total_pages,
    selected.has_next,
  );
  const appendix_pointer = _build_appendix_pointer(has_appendix);

  const ui_markdown =
    banner + selected.page_content + footer + appendix_pointer;
  const llm_content = `> **File Path:** \`${resolve(file_path)}\`\n\n${ui_markdown}`;

  return {
    content: [{ type: "text", text: llm_content }],
    // Include structuredContent for the UI to render the markdown
    structuredContent: {
      markdown: ui_markdown,
      file_path: resolve(file_path),
      title: basename(file_path),
    },
  };
}

export function build_outline_response(
  doc: DocumentObject,
  projected_text: string,
  file_path: string,
  outline_max_level: number = 2,
  outline_verbose: boolean = false,
  paragraph_offsets: Map<any, [number, number]> | null = null,
): ToolResult {
  // Levels outside 1-6 are meaningless (0/negative would render a
  // nonsensical "L1-L0" range label, QA L2). Clamp to the nearest sensible
  // depth, mirroring the Python builder.
  outline_max_level = Math.max(1, Math.min(outline_max_level, 6));

  const [body] = split_structural_appendix(projected_text);
  const pagination_result = paginate(body, "");

  const nodes = extract_outline(
    doc,
    body,
    pagination_result.body_pages,
    pagination_result.body_page_offsets,
    paragraph_offsets,
  );

  const rendered = render_outline_tree(
    nodes,
    outline_max_level,
    outline_verbose,
  );

  const visible_count = nodes.filter(
    (n) => n.level <= outline_max_level,
  ).length;
  const deeper_count = nodes.length - visible_count;
  const deeper_hint =
    deeper_count > 0
      ? ` (${deeper_count} more at deeper levels, raise outline_max_level to see)`
      : "";

  const header = `> **Outline view** — showing ${visible_count} of ${nodes.length} headings (L1-L${outline_max_level}${deeper_hint}) across ${pagination_result.total_pages} page(s). Call \`read_docx\` with \`mode='full'\` and \`page=N\` to read a section.\n\n---\n\n`;
  const ui_markdown = header + rendered;
  const llm_content = `> **File Path:** \`${resolve(file_path)}\`\n\n${ui_markdown}`;

  return {
    content: [{ type: "text", text: llm_content }],
    structuredContent: {
      markdown: ui_markdown,
      file_path: resolve(file_path),
      title: `Outline: ${basename(file_path)}`,
    },
  };
}

export function build_appendix_response(
  text: string,
  page: number,
  file_path: string,
): ToolResult {
  const [, appendix] = split_structural_appendix(text);

  if (!appendix.trim()) {
    const ui_markdown =
      "# Appendix\n\nThis document has no structural appendix (no defined terms, named anchors, or diagnostics detected).";
    const llm_content = `> **File Path:** \`${resolve(file_path)}\`\n\n${ui_markdown}`;
    return {
      content: [{ type: "text", text: llm_content }],
      structuredContent: {
        markdown: ui_markdown,
        file_path: resolve(file_path),
        title: `Appendix: ${basename(file_path)}`,
      },
    };
  }

  const result = paginate(appendix, "");

  if (page < 1 || page > result.total_pages) {
    throw new Error(
      `Appendix page ${page} out of range (appendix has ${result.total_pages} pages).`,
    );
  }

  const selected = result.pages[page - 1];

  let banner = "";
  let footer = "";

  if (selected.total_pages > 1) {
    banner = `> **Appendix page ${selected.page} of ${selected.total_pages}** — structural metadata for this document.\n\n---\n\n`;
    footer = selected.has_next
      ? `\n\n---\n\n> **Continues on appendix page ${selected.page + 1} of ${selected.total_pages}.**`
      : "";
  } else {
    banner =
      "> **Appendix** — structural metadata for this document.\n\n---\n\n";
  }

  const ui_markdown = banner + selected.page_content + footer;
  const llm_content = `> **File Path:** \`${resolve(file_path)}\`\n\n${ui_markdown}`;

  return {
    content: [{ type: "text", text: llm_content }],
    structuredContent: {
      markdown: ui_markdown,
      file_path: resolve(file_path),
      title: `Appendix: ${basename(file_path)}`,
    },
  };
}

export function build_search_response(
  text: string,
  search_query: string,
  search_regex: boolean,
  search_case_sensitive: boolean,
  page: number | string | undefined,
  file_path: string,
): ToolResult {
  const [body] = split_structural_appendix(text);
  const escapeRegExp = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const flags = search_case_sensitive ? "g" : "gi";

  // When the caller asked for a regex but supplied something the engine can't
  // compile (e.g. an unterminated character class `\[`, or an inline-flag group
  // `(?i)...` that JS RegExp rejects), do NOT hard-error and burn the turn.
  // Downgrade to a literal search of the raw string and tell the model, so it
  // can either accept the literal hits or fix its pattern — instead of retrying
  // the same broken regex.
  let regexDowngradedNote = "";
  let regex: RegExp;
  let isUserRegex = false;
  if (search_regex) {
    try {
      regex = new RegExp(search_query, flags);
      isUserRegex = true;
    } catch (e: any) {
      regexDowngradedNote =
        `> **Note:** \`${search_query}\` is not a valid regular expression ` +
        `(${e.message}), so it was searched as literal text instead. ` +
        `If you meant a regex, fix the pattern; if you meant literal text, set \`search_regex\` to false.`;
      regex = new RegExp(escapeRegExp(search_query), flags);
    }
  } else {
    regex = new RegExp(escapeRegExp(search_query), flags);
  }

  // Patterns that blow the matching time budget (catastrophic backtracking,
  // QA 2026-07-17 F5) get the same literal downgrade as invalid patterns —
  // for a read-only search, degraded results beat a hung event loop.
  let allMatches: Array<{ 0: string; index?: number }>;
  if (isUserRegex) {
    try {
      allMatches = userFindAllMatches(search_query, body, flags).map((m) => ({
        0: body.slice(m.start, m.end),
        index: m.start,
      }));
    } catch (e: any) {
      if (!(e instanceof RegexTimeoutError)) throw e;
      regexDowngradedNote =
        `> **Note:** \`${search_query}\` was searched as literal text instead of as ` +
        `a regular expression: ${e.message}`;
      allMatches = Array.from(body.matchAll(new RegExp(escapeRegExp(search_query), flags)));
    }
  } else {
    allMatches = Array.from(body.matchAll(regex));
  }

  // Compute document pagination once — needed for both annotation and filtering.
  const pag_res = paginate(body, "");
  const page_offsets = pag_res.body_page_offsets;
  const total_doc_pages = pag_res.total_pages;

  // Resolve `page` parameter to either "all" or a concrete document-page number.
  // Undefined → "all" (search across the whole document).
  // "all" (case-insensitive) → "all".
  // A positive integer N → filter matches to document page N.
  // Anything else → hard error.
  let filter_doc_page: number | null = null; // null means "all"
  if (page !== undefined && page !== null) {
    const pageStr = String(page).toLowerCase();
    if (pageStr !== "all") {
      const parsed = parseInt(pageStr, 10);
      if (isNaN(parsed) || parsed < 1) {
        throw new Error(
          `Invalid page value: \`${page}\`. Pass a positive integer to restrict the search to that document page, omit \`page\` to search all pages, or pass \`page='all'\` explicitly.`,
        );
      }
      if (parsed > total_doc_pages) {
        throw new Error(
          `Document page ${parsed} is out of range — the document has ${total_doc_pages} page(s). In search mode, \`page\` filters matches to a specific document page; omit it or pass \`page='all'\` to search the whole document.`,
        );
      }
      filter_doc_page = parsed;
    }
  }

  // Helper: which document page does an offset live on?
  const pageOfOffset = (offset: number): number => {
    let p = 1;
    for (let j = 0; j < page_offsets.length; j++) {
      if (offset >= page_offsets[j]) p = j + 1;
      else break;
    }
    return p;
  };

  // Apply the filter (if any), but keep a record of all pages that had hits
  // so we can show a useful summary even when filtered.
  const pagesWithHits = new Set<number>();
  for (const m of allMatches) {
    pagesWithHits.add(pageOfOffset(m.index!));
  }

  const matches =
    filter_doc_page === null
      ? allMatches
      : allMatches.filter((m) => pageOfOffset(m.index!) === filter_doc_page);

  // --- Empty result ---
  if (matches.length === 0) {
    let body_msg: string;
    if (filter_doc_page !== null) {
      if (allMatches.length === 0) {
        body_msg = `> **Search Results** — No matches found for query \`${search_query}\` in \`${basename(file_path)}\`.\n\nVerify your search spelling, or try setting \`search_case_sensitive\` to false or enabling \`search_regex\` if you used pattern wildcards.`;
      } else {
        const hitPages = Array.from(pagesWithHits).sort((a, b) => a - b);
        body_msg = `> **Search Results** — No matches for \`${search_query}\` on document page ${filter_doc_page}.\n\nThe query DOES appear elsewhere in the document (${allMatches.length} match${allMatches.length !== 1 ? "es" : ""} on page${hitPages.length !== 1 ? "s" : ""} ${hitPages.join(", ")}). Omit \`page\` or pass \`page='all'\` to see them.`;
      }
    } else {
      body_msg = `> **Search Results** — No matches found for query \`${search_query}\` in \`${basename(file_path)}\`.\n\nVerify your search spelling, or try setting \`search_case_sensitive\` to false or enabling \`search_regex\` if you used pattern wildcards.`;
    }
    if (regexDowngradedNote) body_msg = `${regexDowngradedNote}\n\n${body_msg}`;
    const llm_content = `> **File Path:** \`${resolve(file_path)}\`\n\n${body_msg}`;
    return {
      content: [{ type: "text", text: llm_content }],
      structuredContent: {
        markdown: body_msg,
        title: `Search: ${basename(file_path)}`,
        file_path: resolve(file_path),
      },
    };
  }

  // --- Build the response ---
  const ui_parts: string[] = [];

  if (filter_doc_page !== null) {
    ui_parts.push(
      `> **Search Results** — Found ${matches.length} match${matches.length !== 1 ? "es" : ""} for \`${search_query}\` on document page ${filter_doc_page} of ${total_doc_pages} in \`${basename(file_path)}\`.`,
    );
    const otherPages = Array.from(pagesWithHits)
      .filter((p) => p !== filter_doc_page)
      .sort((a, b) => a - b);
    if (otherPages.length > 0) {
      ui_parts.push(
        `> Additional matches exist on page${otherPages.length !== 1 ? "s" : ""} ${otherPages.join(", ")} — omit \`page\` or pass \`page='all'\` to see them.`,
      );
    }
  } else {
    ui_parts.push(
      `> **Search Results** — Found ${matches.length} match${matches.length !== 1 ? "es" : ""} for \`${search_query}\` in \`${basename(file_path)}\`.`,
    );
    if (total_doc_pages > 1) {
      // Build a per-page hit distribution: "p1: 3, p3: 1, p7: 12"
      const counts = new Map<number, number>();
      for (const m of allMatches) {
        const p = pageOfOffset(m.index!);
        counts.set(p, (counts.get(p) || 0) + 1);
      }
      const distribution = Array.from(counts.entries())
        .sort((a, b) => a[0] - b[0])
        .map(([p, n]) => `p${p}: ${n}`)
        .join(", ");
      ui_parts.push(
        `> Distribution across ${total_doc_pages} document pages — ${distribution}. Pass \`page=N\` to filter to a specific document page.`,
      );
    }
  }

  // Per-match occurrence counts use the FULL match set, not the filtered one —
  // this gives the LLM accurate global counts even when filtering.
  const occurrences_map: Record<string, number> = {};
  for (const m of allMatches) {
    const matched_str = m[0];
    occurrences_map[matched_str] = (occurrences_map[matched_str] || 0) + 1;
  }

  function get_heading(idx: number, txt: string): string {
    const txtBefore = txt.substring(0, idx);
    const lines = txtBefore.split("\n");
    const path: string[] = [];
    let current_level = 999;

    for (let i = lines.length - 1; i >= 0; i--) {
      const line = lines[i];
      const m = line.match(/^(#{1,6})\s+(.*)/);
      if (m) {
        const level = m[1].length;
        if (level < current_level) {
          let cleanHeading = m[2]
            .replace(/\*\*|__|[*_]/g, "")
            .replace(/\{#[^}]+\}/g, "")
            .trim();
          if (cleanHeading.length > 80) {
            cleanHeading = cleanHeading.substring(0, 80) + "...";
          }
          path.unshift(cleanHeading);
          current_level = level;
          if (level === 1) break;
        }
      }
    }
    return path.join(" > ");
  }

  let i = 1;
  for (const m of matches) {
    const matched_str = m[0];
    const m_start = m.index!;
    const m_end = m_start + matched_str.length;
    const p_num = pageOfOffset(m_start);

    const snippet_start = Math.max(0, m_start - 100);
    const snippet_end = Math.min(body.length, m_end + 100);
    const snippet = emphasizedSnippet(
      body.substring(snippet_start, m_start),
      matched_str,
      body.substring(m_end, snippet_end),
    );

    const snippet_lines = snippet
      .split("\n")
      .filter((line) => line.trim().length > 0)
      .map((line) => `> ${line}`)
      .join("\n");

    ui_parts.push("---");
    ui_parts.push(`### Match ${i} (p${p_num})`);

    const h_path = get_heading(m_start, body);
    if (h_path) {
      ui_parts.push(`**Path:** \`${h_path}\``);
    }

    const count = occurrences_map[matched_str];
    ui_parts.push(snippet_lines);
    ui_parts.push(
      `*Occurrences:* This exact phrasing appears ${count} time${count !== 1 ? "s" : ""} in the document.`,
    );

    i++;
  }

  if (regexDowngradedNote) ui_parts.unshift(regexDowngradedNote);
  const ui_markdown = ui_parts.join("\n\n");
  const llm_content = `> **File Path:** \`${resolve(file_path)}\`\n\n${ui_markdown}`;

  return {
    content: [{ type: "text", text: llm_content }],
    structuredContent: {
      markdown: ui_markdown,
      title: `Search: ${basename(file_path)}`,
      file_path: resolve(file_path),
    },
  };
}
